### PR TITLE
ignore zero-length edges for default epsilon

### PR DESCRIPTION
**Pull request by njwilson23**
possible fix for #5263

### CODE REVIEW

File: scipy/interpolate/rbf.py [hunk 1]
```diff
@@ -200,11 +200,13 @@ def __init__(self, *args, **kwargs):
...
         self.epsilon = kwargs.pop('epsilon', None)
         if self.epsilon is None:
-            # default epsilon is the "the average distance between nodes"
+            # default epsilon is the "the average distance between nodes" based
+            # on a bounding hypercube
             dim = self.xi.shape[0]
             ximax = np.amax(self.xi, axis=1)
             ximin = np.amin(self.xi, axis=1)
-            self.epsilon = np.power(np.prod(ximax - ximin)/self.N, 1.0/dim)
+            edges = np.array([a for a in ximax-ximin if a != 0.0])
```
Comment (ev-br) [inline scipy/interpolate/rbf.py]: this is probably better written as e.g. `edges = ximax - ximin;  edges = edges[np.nonzero(edges)]`.

Comment (njwilson23) [inline scipy/interpolate/rbf.py]: Got it. Also seems to make sense to subtract the zero edges from `dim`.

**Comment by ev-br** [comment]
Thanks! 

Can you add a regression test as well? An example you reported in gh-5263 would be good. Also, a test with d=3 would be nice bonus to check that implementation does not assume 1d or 2d.

What about stability. How does your fix (and the master branch) fare if the points are *nearly* collinear?

**Comment by ev-br** [comment]
The PR which last changed the value of default epsilon: gh-4643.
cc @jjgreen

**Comment by njwilson23** [comment]
I'll take a look at this and add tests. Playing around, nearly collinear points give small epsilon (as expected), but what this does to the interpolation seems to depend a lot on the shape function used. The stability properties here are a bit out of my depth.

**Comment by ev-br** [comment]
Maybe just try to use collinear points + small noise and check that the result is sensible? Maybe something like this can even turned into a test.

I'd also ping  @arokem: you seemed to have a stake in the rbf implementation here?

**Comment by njwilson23** [comment]
I wrote a simple regression test in c3a10c4c71aa068ae1b9cddafd7a8a6d49b0e29f for the perfectly collinear case.

When points are nearly collinear and epsilon is small, the `multiquadric` basis returns reasonable results (turns into linear interpolation). The `gaussian` and `inverse` shapefunctions break down because the basis becomes very "narrow".

I have not yet turned this into another test because I'm not sure what would actually be tested.

**Comment by arokem** [comment]
Thanks for pinging me about this. I think this will be fine for our use-case - we've decided to go with determining epsilon before calling this, so we will not be affected by this change.

File: scipy/interpolate/tests/test_rbf.py [hunk 2]
```diff
@@ -146,5 +146,15 @@ def test_rbf_epsilon_none():
...
 
 
+def test_rbf_epsilon_none_collinear():
+    # Check that collinear points in one dimension doesn't cause an error
+    # due to epsilon = 0
+    x = [1, 2, 3]
+    y = [4, 4, 4]
+    z = [5, 6, 7]
+    rbf = Rbf(x, y, z, epsilon=None)
+    assert_almost_equal(rbf.epsilon, 2.0/3)
```
Comment (ev-br) [inline scipy/interpolate/tests/test_rbf.py]: This is too strict IMO. 2/3 is an implementation detail and I don't think we should insist on this particular estimate of the default epsilon. I'd just test that epsion is non-zero, e.g., `assert_(rbf.epsilon > 0)`.

**Comment by ev-br** [comment]
Thank you @arokem 

OK, it seems to me that (i) our RBF implementation can really use a face-lift, (ii) users are best off managing their values of epsilon themselves,  (iii) we should provide a sensible default if only for backwards compat. There are likely better ways of doing it, but using a bounding hypercube is probably an OK default, and this PR is a bugfix for an otherwise OK behavior. In view of (ii) we can probably leave an issue with nearly collinear points. Therefore, I'm +1 for merging it and backporting to 0.16.x (assuming 0.16.1 is going to happen).

One last minor thing to do before merging this PR: can you squash the four commits into one @njwilson23? Ideally, the commit message would start with BUG (as per http://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html).

**Comment by njwilson23** [comment]
Point about the test taken. I've also squashed the commits and pre-pended BUG to the commit message.

**Comment by ev-br** [comment]
Merging, thank you @njwilson23. Keep them coming!

**Pull request merged**